File: framework/test/04/leet.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class TopVotedCandidate:

    def __init__(self, persons: List[int], times: List[int]):
        self.N = len(times)
        self.times = times
        self.winner = [-1] * self.N
        self.initialize(persons)

    def initialize(self, persons):
        counter = {}
        counter[-1] = 0
        win = -1
        vote = 0
        for i in range(self.N):
            v = counter.get(persons[i], 0) + 1
            if v >= vote:
                win = persons[i]
                vote = v
            counter[persons[i]] = v
            self.winner[i] = win

    def q(self, t: int) -> int:
        if t >= self.times[-1]:
            return self.winner[-1]
        low, high = 0, self.N-1
        while low < high:
            mid = (low + high) // 2
            if t <= self.times[mid]:
                high = mid
            else:
                low = mid + 1
        return self.winner[low] if t == self.times[low] else self.winner[low-1]

class Params:
    def __init__(self):
        self.persons = None
        self.times = None
        self.qs = None

class Solution:
    def doTest(self, commands: List[str], params: Params) -> List[int]:
        res = []
        tvc = None
        for i, cmd in enumerate(commands):
            if cmd == 'TopVotedCandidate':
                tvc = TopVotedCandidate(params.persons, params.times)
                res.append(None)
            else:
                res.append(tvc.q(params.qs[i]))
        return res

def params_parser(p):
    params = Params()
    params.persons = p[0][0]
    params.times = p[0][1]
    params.qs = [-1] * len(p)
    for i in range(1, len(p)):
        params.qs[i] = p[i][0]
    return params

if __name__ == '__main__':
    init_logging()
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().doTest)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    work.setArgumentParser(1, params_parser)
    # work.setArgumentParser(index, (a) => b)
    # work.resultSerializer = (r) => s
    # work.resultParser = (s) => r
    work.run()

File: framework/test/04/leet.go
package main

import "encoding/json"
import "fmt"

import _ "soda/leetcode"
import "soda/unittest"
import "soda/util"

type TopVotedCandidate struct {
    N int
    times []int
    winner []int
}

func (this *TopVotedCandidate) initialize(persons []int) {
    counter := make([]int, this.N+1)
    win := 0
    for i := 0; i < this.N; i++ {
        counter[persons[i]]++
        if counter[persons[i]] >= counter[win] {
            win = persons[i]
        }
        this.winner[i] = win
    }
}

func Constructor(persons []int, times []int) TopVotedCandidate {
    var tvc TopVotedCandidate
    tvc.N = len(persons)
    tvc.times = times
    tvc.winner = make([]int, tvc.N)
    tvc.initialize(persons)
    return tvc
}


func (this *TopVotedCandidate) Q(t int) int {
    if t >= this.times[len(this.times)-1] {
        return this.winner[len(this.times)-1]
    }
    low, high := 0, this.N - 1
    for low < high {
        mid := (low + high) / 2
        if t <= this.times[mid] {
            high = mid
        } else {
            low = mid + 1
        }
    }
    if t == this.times[low] {
        return this.winner[low]
    } else {
        return this.winner[low-1]
    }
}

type Params struct {
    persons []int
    times []int
    qs []int
}

// func (this *Params) UnmarshalJSON(data []byte) error {
//     var temp []json.RawMessage
//     if err := json.Unmarshal(data, &temp); err != nil {
//         return err
//     }
//     var pt [][]int
//     if err := json.Unmarshal(temp[0], &pt); err != nil {
//         return err
//     }
//     this.persons = pt[0]
//     this.times = pt[1]
//     this.qs = append(this.qs, -1)
//     for i := 1; i < len(temp); i++ {
//         var values []int
//         if err := json.Unmarshal(temp[i], &values); err != nil {
//             return err
//         }
//         this.qs = append(this.qs, values[0])
//     }
//     return nil
// }

func doTest(commands []string, params Params) util.OptionalIntSlice {
    var tvc TopVotedCandidate
    res := make(util.OptionalIntSlice, len(commands))
    for i, cmd := range commands {
        var optInt util.OptionalInt
        if cmd == "TopVotedCandidate" {
            tvc = Constructor(params.persons, params.times)
        } else {
            r := tvc.Q(params.qs[i])
            optInt.Set(r)
        }
        res[i] = optInt
    }
    return res
}

func parse_params(temp []json.RawMessage) Params {
    var this Params
    var pt [][]int
    if err := json.Unmarshal(temp[0], &pt); err != nil {
        panic(fmt.Sprintf("json unmarshal error: %s\n", err))
    }
    this.persons = pt[0]
    this.times = pt[1]
    this.qs = append(this.qs, -1)
    for i := 1; i < len(temp); i++ {
        var values []int
        if err := json.Unmarshal(temp[i], &values); err != nil {
            panic(fmt.Sprintf("json unmarshal error: %s\n", err))
        }
        this.qs = append(this.qs, values[0])
    }
    return this
}

func main() {
    work := unittest.CreateWork(doTest)
    // work.SetValidator(func(e,r)bool)
    work.CompareSerial = true
    work.SetArgParser(1, parse_params)
    // work.SetResultParser(func(s)w)
    // work.SetResultSerializer(func(w)s)
    work.Run()
}
